wakuv2/common/helpers.go

```diff
@@ -110,3 +114,124 @@ func ValidateDataIntegrity(k []byte, expectedSize int) bool {
 	}
 	return true
 }
+
+func ParseDialErrors(errMsg string) []DialError {
+	// Regular expression to match the array of failed dial attempts
+	re := regexp.MustCompile(`all dials failed\n((?:\s*\*\s*\[.*\].*\n?)+)`)
+
+	match := re.FindStringSubmatch(errMsg)
+	if len(match) < 2 {
+		return nil
+	}
+
+	// Split the matched string into individual dial attempts
+	dialAttempts := strings.Split(strings.TrimSpace(match[1]), "\n")
+
+	// Regular expression to extract multiaddr and error message
+	reAttempt := regexp.MustCompile(`\[(.*?)\]\s*(.*)`)
+
+	var dialErrors []DialError
+	for _, attempt := range dialAttempts {
+		attempt = strings.TrimSpace(strings.Trim(attempt, "* "))
+		matches := reAttempt.FindStringSubmatch(attempt)
+		if len(matches) == 3 {
+			errMsg := strings.TrimSpace(matches[2])
+			ma, err := multiaddr.NewMultiaddr(matches[1])
+			if err != nil {
+				continue
+			}
+			protocols := ma.Protocols()
+			protocolsStr := "/"
+			for i, protocol := range protocols {
+				protocolsStr += protocol.Name
+				if i < len(protocols)-1 {
+					protocolsStr += "/"
+				}
+			}
+			dialErrors = append(dialErrors, DialError{
+				Protocols: protocolsStr,
+				MultiAddr: matches[1],
+				ErrMsg:    errMsg,
+				ErrType:   CategorizeDialError(errMsg),
+			})
+		}
+	}
+
+	return dialErrors
+}
+
+// DialErrorType represents the type of dial error
+type DialErrorType int
+
+const (
+	ErrorUnknown DialErrorType = iota
+	ErrorIOTimeout
+	ErrorConnectionRefused
+	ErrorRelayCircuitFailed
+	ErrorRelayNoReservation
+	ErrorSecurityNegotiationFailed
+	ErrorConcurrentDialSucceeded
+	ErrorConcurrentDialFailed
+	ErrorConnectionsPerIPLimitExceeded
+	ErrorStreamReset
+	ErrorRelayResourceLimitExceeded
+	ErrorOpeningHopStreamToRelay
+	ErrorDialBackoff
+)
+
+func (det DialErrorType) String() string {
+	return [...]string{
+		"Unknown",
+		"I/O Timeout",
+		"Connection Refused",
+		"Relay Circuit Failed",
+		"Relay No Reservation",
+		"Security Negotiation Failed",
+		"Concurrent Dial Succeeded",
+		"Concurrent Dial Failed",
+		"Connections Per IP Limit Exceeded",
+		"Stream Reset",
+		"Relay Resource Limit Exceeded",
+		"Error Opening Hop Stream to Relay",
+		"Dial Backoff",
+	}[det]
```

Comment:
Though I like the hack, I think `map[string]string` is a more reliable solution in terms of potential changes. And the performance should remain the same.